### ISSUE TITLE
[typedef] False positive errors on deep destructuring

### ISSUE THREAD

**Issue by gultyaev** [typescript-eslint/typescript-eslint#4505]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/typedef": [
      "warn",
          {
            "arrayDestructuring": true,
            "arrowParameter": true,
            "memberVariableDeclaration": true,
            "objectDestructuring": true,
            "parameter": true,
            "propertyDeclaration": true,
            "variableDeclaration": true,
            "variableDeclarationIgnoreFunction": true
          }
    ]
  }
}
```

```TS
const {
      user: { name }, // <-- It complains about type annotation for { name }
      email,
    }: User = await this.getUser();
```

**Expected Result**

Should not print any warnings

**Actual Result**

Shows a warning (because of that I cannot switch the rule to "error" instead of "warn")

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `^5.9.1` |
| `@typescript-eslint/parser`        | `^5.9.1` |
| `TypeScript`                       | `~4.3.5` |
| `ESLint`                           | `^8.6.0` |
| `node`                             | `16.13.1` |

**Comment by JoshuaKGoldberg** [comment]
Thanks for the spot @gultyaev! Agreed, this is a bug we'd like a fix for.

Btw it's generally helpful to have a _full_ reproduction in the issue to make it easier for us to figure out what's going on. Meaning, I should be able to copy & paste it or open an link without messing around on my end. This source snippet shows the bad behavior:

```ts
interface User {
    details: { firstName: string };
    email: string;
}

function getUser(): User {
    return {
        details: { firstName: '' },
        email: '',
    }
}

const {
    details: { firstName }, // <-- It complains about type annotation for { name }
    email,
}: User = getUser();
```

https://typescript-eslint.io/play/#ts=4.5.2&sourceType=module&code=JYOwLgpgTgZghgYwgAgKoGdrIN4FgBQyRyAJhGHMADboBcOyMwU6YAcnALYT2tSgBzZAF8A3AWLIInSlV5h+IAePzCCBGAFcQCMMAD2IZAPIZoACgCU9M1BwTiUcpqhG8hScTIVqdBkxZ2Lh5kAHJQkQAaB08pGWp6cOiPIjVVdXwEQ1Z7FNJyWT9sRmZWDm4o5AB6KuQAHgBaBuQASTBkLM4ABypKEHRkOAAjfU12sABPLpQ4EBB9Cj1DRn07YpBgkRi42WThG0w7AF5jU0OrFSA&rules=N4IgAgLgngDgpgZwMYCcCWMIFpEBs0B2EA9NPACZwBmIAXANogDuAhigSADShsotQARRBBQBXJBFHoCAczojRcTiF4B7JgAU2LALZwIcFPLFKQenQCNDANTZoWF3HCFJc2iGlUdaC06osAVnASQggKElKEcj4myjDaegZGMYpxKKrwKNAubnweXsapIABudg5OOe6e3r7KpejlzsG5LPkEAJIyBKoocABiogQS1YVwAL4AumNAA&tsConfig=N4XyA
